salt/cloud/clouds/saltify.py

```diff
@@ -47,30 +54,150 @@ def __virtual__():
+
     return {}
 
 
-def list_nodes_select():
+def avail_images(call=None):
+    '''This function returns a list of images available for this cloud provider.
+     saltify will return a list of profiles.
+     salt-cloud --list-images my-cloud-provider
     '''
-    Because this module is not specific to any cloud providers, there will be
-    no nodes to list.
+    vm_ = get_configured_provider()
+    return {'Profiles': [profile for profile in vm_['profiles']]}
+
+
+def avail_sizes(call=None):
+    '''
+    This function returns a list of sizes available for this cloud provider.
+    salt-cloud --list-sizes my-cloud-provider
+    [ saltify always returns an empty dictionary ]
     '''
     return {}
 
 
+def list_nodes(call=None):
+    '''
+    List the nodes which have salt-cloud:driver:saltify grains.
+    salt-cloud -Q
+    '''
+    nodes = _list_nodes_full(call)
+    return _build_required_items(nodes)
+
+
+def _build_required_items(nodes):
+    ret = {}
+    for name, grains in nodes.items():
+        if grains:
+            private_ips = []
+            public_ips = []
+            ips = grains['ipv4'] + grains['ipv6']
+            for adrs in ips:
+                ip_ = ipaddress.ip_address(adrs)
+                if not ip_.is_loopback:
+                    if ip_.is_private:
+                        private_ips.append(adrs)
+                    else:
+                        public_ips.append(adrs)
+
+            ret[name] = {
+                'id': grains['id'],
+                'image': grains['salt-cloud']['profile'],
+                'private_ips': private_ips,
+                'public_ips': public_ips,
+                'size': '',
+                'state': 'running'
+            }
+
+    return ret
+
+
+def list_nodes_full(call=None):
+    '''
+    List the nodes, ask all 'saltify' minions, return dict of grains (enhanced).
+    '''
+    ret = _list_nodes_full(call)
+
+    for key, grains in ret.items():  # clean up some hyperverbose grains -- everything is too much
+        try:
+            del grains['cpu_flags'], grains['disks'], grains['pythonpath'], grains['dns'], grains['gpus']
+        except KeyError:
+            pass  # ignore absence of things we are eliminating
+        except TypeError:
+            del ret[key]  # eliminate all reference to unexpected (None) values.
+
+    reqs = _build_required_items(ret)
+    for name in ret:
+        ret[name].update(reqs[name])
+    return ret
+
+
+def _list_nodes_full(call=None):
+    '''
+    List the nodes, ask all 'saltify' minions, return dict of grains.
+    '''
+    local = salt.netapi.NetapiClient(__opts__)
+    cmd = {'client': 'local',
+           'tgt': 'salt-cloud:driver:saltify',
+           'fun': 'grains.items',
+           'arg': '',
+           'tgt_type': 'grain',
+           }
+    cmd.update(_get_connection_info())
+
+    return local.run(cmd)
+
+
+def list_nodes_select(call=None):
+    ''' Return a list of the minions that have salt-cloud grains, with
```

Comment:
There are a couple of other doc lines that are set-up this way. Can you go through and fix the rest as well?